### PR TITLE
Optimizations for HAProxy reloads

### PR DESCRIPTION
**Pull request by jmencak**
@jeremyeder could you please take a look?

/cc @ahardin-rh

### CODE REVIEW

File: scaling_performance/routing_optimization.adoc [hunk 1]
```diff
@@ -125,3 +125,19 @@ accepted by load balancers provided by many public cloud providers. However,
...
+reload intervals can cause instantiation of many HAProxy processes.
+These processes need to handle old connections which were started
+before the HAProxy configuration reload.  Large number of these processes is
+undesirable, as it will exert unnecessary load on the system and may
+lead to issues, such as OoM conditions.
+
+xref:../architecture/topics/router_environment_variables.adoc#[Router environment variables] affecting this
+behaviour are `ROUTER_DEFAULT_TUNNEL_TIMEOUT`, `ROUTER_DEFAULT_CLIENT_TIMEOUT`, 
+`ROUTER_DEFAULT_SERVER_TIMEOUT` and `RELOAD_INTERVAL` in particular.  It is currently recommended to 
+set `RELOAD_INTERVAL` to 15s.
```
[the remark this answers is not in the document]
Comment (jmencak) [reply scaling_performance/routing_optimization.adoc]: 1) Default reload interval is 5s.
2) 15s comes from what Ben recommended in https://bugzilla.redhat.com/show_bug.cgi?id=1471899 , but that was to address an immediate BZ issue, which should now be fixed in other ways.  Perhaps I should re-formulate this or remove this completely as this is trying to fix another problem.  15s could make the number of process 3x lower, in theory.
3) Not sure about this.  The docs are for 3.8+ and the the ~3000 route issue I was seeing should be fixed by https://github.com/openshift/origin/pull/17049, which already merged.

[the remark this answers is not in the document]
Comment (jeremyeder) [reply scaling_performance/routing_optimization.adoc]: Sounds like we need to re-test after #17049 and them re-approach this particular documentation.  Unless you wanted to target this at "3.7 and earlier" versions of the docs only?

[the remark this answers is not in the document]
Comment (jmencak) [reply scaling_performance/routing_optimization.adoc]: This section has nothing to do with #17049 or BZ1471899 other than the fact increasing RELOAD_INTERVAL can alleviate the problems seen by BZ1471899 for versions 3.6 and earlier.  I didn't write the section for that purpose.  This section has everything to do with the inherent incapability of HAProxy to reload configuration without forking another process while serving (old and new) connections.  BZ1471899 and  #17049 was retested by QA, but I can retest, not a big deal.

File: scaling_performance/routing_optimization.adoc [hunk 2]
```diff
@@ -125,3 +125,18 @@ accepted by load balancers provided by many public cloud providers. However,
 this affects the total memory use, especially when large numbers of connections
 are open. With very large numbers of open connections, the memory usage will be
 nearly proportionate to the increase of this tunable parameter.
+
+[[optimizations-for-haproxy-reloads]]
+==== Optimizations for HAProxy reloads
```
Comment (ahardin-rh) [inline scaling_performance/routing_optimization.adoc]: Reloads

File: scaling_performance/routing_optimization.adoc [hunk 3]
```diff
@@ -125,3 +125,18 @@ accepted by load balancers provided by many public cloud providers. However,
...
+
+Long-lasting connections, such as WebSocket connections, combined with
+long client/server HAProxy timeouts and short HAProxy
+reload intervals can cause instantiation of many HAProxy processes.
```
Comment (ahardin-rh) [inline scaling_performance/routing_optimization.adoc]: intervals,

File: scaling_performance/routing_optimization.adoc [hunk 4]
```diff
@@ -125,3 +125,18 @@ accepted by load balancers provided by many public cloud providers. However,
...
+These processes need to handle old connections which were started
```
Comment (ahardin-rh) [inline scaling_performance/routing_optimization.adoc]: processes must handle old connections, which

File: scaling_performance/routing_optimization.adoc [hunk 5]
```diff
@@ -125,3 +125,18 @@ accepted by load balancers provided by many public cloud providers. However,
...
+before the HAProxy configuration reload.  Large number of these processes is
```
Comment (ahardin-rh) [inline scaling_performance/routing_optimization.adoc]: extra space before large

A large number of...

File: scaling_performance/routing_optimization.adoc [hunk 6]
```diff
@@ -125,3 +125,18 @@ accepted by load balancers provided by many public cloud providers. However,
...
+undesirable, as it will exert unnecessary load on the system and may
+lead to issues, such as OoM conditions.
```
Comment (ahardin-rh) [inline scaling_performance/routing_optimization.adoc]: s/OoM/out of memory

File: scaling_performance/routing_optimization.adoc [hunk 7]
```diff
@@ -125,3 +125,18 @@ accepted by load balancers provided by many public cloud providers. However,
...
+
+[[optimizations-for-haproxy-reloads]]
+==== Optimizations for HAProxy reloads
+
+Long-lasting connections, such as WebSocket connections, combined with
+long client/server HAProxy timeouts and short HAProxy
+reload intervals can cause instantiation of many HAProxy processes.
+These processes need to handle old connections which were started
+before the HAProxy configuration reload.  Large number of these processes is
+undesirable, as it will exert unnecessary load on the system and may
```
Comment (ahardin-rh) [inline scaling_performance/routing_optimization.adoc]: s/may/can

File: scaling_performance/routing_optimization.adoc [hunk 8]
```diff
@@ -125,3 +125,18 @@ accepted by load balancers provided by many public cloud providers. However,
...
+lead to issues, such as OoM conditions.
+
+xref:../architecture/topics/router_environment_variables.adoc#[Router environment variables] affecting this
+behaviour are `ROUTER_DEFAULT_TUNNEL_TIMEOUT`, `ROUTER_DEFAULT_CLIENT_TIMEOUT`, 
+`ROUTER_DEFAULT_SERVER_TIMEOUT` and `RELOAD_INTERVAL` in particular.
```
Comment (ahardin-rh) [inline scaling_performance/routing_optimization.adoc]: `ROUTER_DEFAULT_SERVER_TIMEOUT`, and `RELOAD_INTERVAL`.

File: scaling_performance/routing_optimization.adoc [hunk 9]
```diff
@@ -125,3 +125,18 @@ accepted by load balancers provided by many public cloud providers. However,
...
+Long-lasting connections, such as WebSocket connections, combined with
+long client/server HAProxy timeouts and short HAProxy
+reload intervals can cause instantiation of many HAProxy processes.
+These processes need to handle old connections which were started
+before the HAProxy configuration reload.  Large number of these processes is
+undesirable, as it will exert unnecessary load on the system and may
+lead to issues, such as OoM conditions.
+
+xref:../architecture/topics/router_environment_variables.adoc#[Router environment variables] affecting this
+behaviour are `ROUTER_DEFAULT_TUNNEL_TIMEOUT`, `ROUTER_DEFAULT_CLIENT_TIMEOUT`,
```
Comment (ahardin-rh) [inline scaling_performance/routing_optimization.adoc]: s/behaviour/behavior

**Comment by ahardin-rh** [comment]
@jmencak Just some nits from me. Thanks for this! Also, can you please confirm that this is targeting 3.9? Is there an associated Trello card? Thanks again!

**Comment by jmencak** [comment]
Yes, targetting 3.9.  No trello card, simple change.

**Comment by ahardin-rh** [comment]
Latest changes LGTM. Thanks!

**Comment by ahardin-rh** [comment]
[rev_history]
|xref:../scaling_performance/routing_optimization.adoc#scaling-performance-routing-optimization[Routing Optimization]
|Added a new xref:../scaling_performance/routing_optimization.adoc#optimizations-for-haproxy-reloads[Optimizations for HAProxy Reloads] section.
%

**Pull request merged**